Reimplement preallocate for hashed/sparse_hashed dictionaries

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Preallocate support for hashed/sparse_hashed dictionaries

Detailed description / Documentation draft:
It was initially implemented in #15454, but was reverted in #21948 (due
to higher memory usage).

This implementation differs from the initial, since now there is
separate attribute to enable preallocation, before it was done
automatically, but this has problems with duplicates in the source.

Plus this implementation does not uses dynamic_cast, instead it extends
IDictionarySource interface.

### benchmarks

| name | type | duration | elements | bytes | prealloc win
|:-|:-|-:|:-|:-|:-|
| graph_hashed | Hashed | 702.099 | 2.28 billion | 128.00 GiB |
| graph_hashed_preallocate | Hashed | 509.606 | 2.28 billion | 128.00 GiB | **27%**
| graph_sparsehashed | SparseHashed | 3562.452 | 2.28 billion | 21.21 GiB |
| graph_sparsehashed_preallocate | SparseHashed | 2747.515 | 2.28 billion | 21.21 GiB | **30%**

Cc: @kitaisreal

azat committed:
Reimplement preallocate for hashed/sparse_hashed dictionaries

It was initially implemented in #15454, but was reverted in #21948 (due
to higher memory usage).

This implementation differs from the initial, since now there is
separate attribute to enable preallocation, before it was done
automatically, but this has problems with duplicates in the source.

Plus this implementation does not uses dynamic_cast, instead it extends
IDictionarySource interface.

azat committed:
Drop unused getDictionaryConfigurationFromAST.h

azat committed:
Attach background thread for dictionary reload to the reload query

kitaisreal commented:
@azat please provide some benchmark where there will be significant performance improvements.

alexey-milovidov commented:
Ok, but... is it worth adding a separate parameter (that almost no one will use - dead code, difficult to test and support)?

azat commented:
>@azat please provide some benchmark where there will be significant performance improvements.

@kitaisreal There are some numbers in the initial PR (#15454):

```
$ ninja bench-sparse_hash-run
[1/1] cd /src/ch/hashtable-bench/.cmake && ...ch/hashtable-bench/.cmake/bench-sparse_hash
sparse_hash/insert: 7.574 <!--
sparse_hash/find  : 2.14426
sparse_hash/maxrss: 174MiB
sparse_hash/time:   9710.51 msec (user+sys)

$ time ninja bench-sparse_hash-preallocate-run
[1/1] cd /src/ch/hashtable-bench/.cmake && ...-bench/.cmake/bench-sparse_hash-preallocate
sparse_hash/insert: 5.0522 <!--
sparse_hash/find  : 2.14024
sparse_hash/maxrss: 174MiB
sparse_hash/time:   7192.06 msec (user+sys)
```

Plus some numbers from @KochetovNicolai - https://github.com/ClickHouse/ClickHouse/pull/15454#issuecomment-704309099

azat commented:
@alexey-milovidov 

>Ok, but... is it worth adding a separate parameter (that almost no one will use - dead code, difficult to test and support)?

Making it by default increases memory usage (that's why it was reverted in #21948)

>Also, it speeds up only loading of dictionaries. But it is what is mostly irrelevant for performance. Lookup performance is the most important and I don't care if loading is 1.5 times faster or slower.

Well, when loading time is ~1 hour (due to lots of elements), making it 33% faster is significant

azat committed:
Decrease log level in IPAddressDictionary for duplicated records

After RELOAD DICTIONARY had been changed to capture the logs (by
attaching query's thread group) the 01852_dictionary_found_rate_long
fails because of this warning.
But it does not looks like it should be a warning, so change to trace.

azat committed:
Fix google-default-arguments in IDictionarySource with loadAllWithSizeHint()

alexey-milovidov commented:
How do you have one hour loading time?
If dictionary occupies 100 GB, it means just about 30 MB/sec - very slow even for google::sparse_hash_map.

alexey-milovidov commented:
For large hash table the difference in loading time will probably be even lower, because resizes will take less relative amount of time.

alexey-milovidov commented:
What numbers do you have in your production?

alexey-milovidov commented on src/Dictionaries/HashedDictionary.cpp:
```diff
@@ -359,6 +357,8 @@ void HashedDictionary<dictionary_key_type, sparse>::createAttributes()
 template <DictionaryKeyType dictionary_key_type, bool sparse>
 void HashedDictionary<dictionary_key_type, sparse>::updateData()
 {
+    /// NOTE: updateDate() does not preallocation since it may increase memory usage.
```
updateDate -> updateData

azat commented:
>How do you have one hour loading time?
>If dictionary occupies 100 GB, it means just about 30 MB/sec - very slow even for google::sparse_hash_map.

I have even smaller for sparse_hash_map, ~6MB/s (for 2.28 billion of rows)
But this depends from the item size, size of value is 2 bytes, so it is  ~0.64e6 keys per second (w/o preallocation)

And for clickhouse builtin hashtable, ~30MB/s w/o preallocation

>What numbers do you have in your production?

| name | type | duration | elements | bytes | prealloc win
|:-|:-|-:|:-|:-|:-|
| webgraph_hashed | Hashed | 702.099 | 2.28 billion | 128.00 GiB |
| webgraph_hashed_preallocate | Hashed | 509.606 | 2.28 billion | 128.00 GiB | **27%**
| webgraph_sparsehashed | SparseHashed | 3562.452 | 2.28 billion | 21.21 GiB |
| webgraph_sparsehashed_preallocate | SparseHashed | 2747.515 | 2.28 billion | 21.21 GiB | **30%**

>For large hash table the difference in loading time will probably be even lower, because resizes will take less relative amount of time.

The benefit should be greater, not smaller.

alexey-milovidov commented:
Let's try to replace google::sparse_hash_map to something else instead of merging this PR.
Maybe robin hood hash table with 95% fill factor will be good for this use case.

azat commented:
>Let's try to replace google::sparse_hash_map to something else instead of merging this PR.

When I added sparsehash (back in #6894), the main goal was to keep memory footprint as low as possible.
Also note that prellocation make difference even for the builtin hashtable (same 30% percent +-).

> Maybe robin hood hash table with 95% fill factor will be good for this use case.

I've checked robin hood hash table in https://github.com/ClickHouse/ClickHouse/pull/6894#issuecomment-532570298 but it shows greater memory usage.

I've checked robin hood with 95 max load factor and does not differs from the default max load factor 80:

```
[56/63] cd /src/ch/hashtable-bench/.cmake && /src/ch/hashtable-bench/.cmake/bench-robin_hood-max95
robin_hood_95/insert: 4.43618
robin_hood_95/find  : 0.739763
robin_hood_95/maxrss: 411MiB 
robin_hood_95/time:   4557.99 msec (user+sys)

[53/63] cd /src/ch/hashtable-bench/.cmake && /src/ch/hashtable-bench/.cmake/bench-robin_hood-max95-preallocate
robin_hood_95/insert: 2.45228
robin_hood_95/find  : 1.47531
robin_hood_95/maxrss: 275MiB 
robin_hood_95/time:   3028.28 msec (user+sys)

[47/63] cd /src/ch/hashtable-bench/.cmake && /src/ch/hashtable-bench/.cmake/bench-robin_hood
robin_hood/find  : 1.48552
robin_hood/maxrss: 411MiB
robin_hood/time:   3694.86 msec (user+sys)

[52/63] cd /src/ch/hashtable-bench/.cmake && /src/ch/hashtable-bench/.cmake/bench-robin_hood-preallocate
robin_hood/insert: 1.80116
robin_hood/find  : 1.62977
robin_hood/maxrss: 275MiB
robin_hood/time:   2998.08 msec (user+sys)

[60/63] cd /src/ch/hashtable-bench/.cmake && /src/ch/hashtable-bench/.cmake/bench-sparse_hash
sparse_hash/insert: 10.9367
sparse_hash/find  : 1.76198
sparse_hash/maxrss: 174MiB
sparse_hash/time:   11522.4 msec (user+sys)

[58/63] cd /src/ch/hashtable-bench/.cmake && /src/ch/hashtable-bench/.cmake/bench-sparse_hash-preallocate
sparse_hash/insert: 7.89093
sparse_hash/find  : 2.56655
sparse_hash/maxrss: 174MiB
sparse_hash/time:   9549.78 msec (user+sys)
```

Also as you can see robin hood hash table takes less memory only with preallocation.

alexey-milovidov commented:
@kitaisreal I'm ok adding this flag.

gyuton commented:
Internal documentation ticket: DOCSUP-9288.
